empirical/GMM_models/McVerry_2006_Sa.py

```
@@ -68,53 +68,66 @@ def McVerry_2006_Sa(siteprop, faultprop, im=None, period=None):
-            sigma_SA = [np.interp(np.log(period), x, sigma_SA_lh[i]) for i in range(3)]
+        periods = [0]
+
+    try:
+        periods[0]
+    except (TypeError, IndexError):
+        periods = [periods]
+
+    results = []
+    max_period = period_list[-1]
+    for period in periods:
+        if period > max_period:
+            McVerry_max = McVerry06(siteprop, faultprop, max_period)
+            median_max = McVerry_max[0]
+            median = median_max * (max_period / period) ** 2
+            result = (median, McVerry_max[1])
+
+        # interpolate between periods if neccesary
+        elif not np.isclose(period_list, period, atol=0.0001).any():
+            p = np.argmin(period_list < period)
+            assert p > 0
+            T_low = period_list[p - 1]
+            T_hi = period_list[p]
+
+            McVerry_low = McVerry06(siteprop, faultprop, period=T_low)
+            McVerry_high = McVerry06(siteprop, faultprop, period=T_hi)
+
+            result = interpolate_to_closest(period, T_hi, T_low, McVerry_high, McVerry_low)
         else:
-            # linear interpolation
-            x = T_low, T_hi
+            result = McVerry06(siteprop, faultprop, period=period)
 
-            SA = np.interp(period, x, (SA_low, SA_high))
-            sigma_SA = [np.interp(period, x, sigma_SA_lh[i]) for i in range(3)]
+        results.append(result)
 
-        return SA, sigma_SA
+    if len(periods) == 1:
+        results = results[0]
 
-    # Identify the period
-    i = np.argmin(np.abs(periods - period))
+    return results
+
+
+def McVerry06(siteprop, faultprop, period):
+    M = faultprop.Mw
+    R = siteprop.Rrup
 
+    # Identify the period
+    i = np.argmin(np.abs(period_list - period))
     # site class
-    delC = int(siteprop.siteclass.value == "C")
-    delD = int(siteprop.siteclass.value == "D")
+    delC, delD = 0, 0
+    if hasattr(siteprop, "siteclass") and siteprop.siteclass is not None:
```

One quick comment. I'm worried that at some point a siteclass that isn't A, B, C or D will leak in here and go through the top section of this if statement. Can this be made so it's something like (not correct syntax but you'll get what I mean):

siteprop.siteclass is in ['A', 'B', 'C', 'D']

instead of 

siteprop.siteclass is not None

Or would you consider that to be the user's problem?

I'm also thinking about what to do with siteclass E (for which the McVerry model doesn't exist for). The code in the commit would default it to siteclass A/B. If we pass in both siteclass and vs30, then the proposed change above would likely make it siteclass D (which is closest and imo the best). Keen to hear your thoughts anyway.

---

Siteclass E definately was something I've forgotten to ask you about. I agree it should probably be considered D.

Regarding siteclass, as someone who sets it they should set it using the enum, I don't think that anyone would be calling the empirical model (with a siteclass) without understanding the code within- so most uses of it should be okay.
I think it should be okay as is, but happy to change it to what you proposed.

---

Hmm... Fair point. I think I'd be more comfortable with the change to what I propose above. No real downsides to it and it "resolves" the siteclass E issue this way. Would be good if you can make that change and then that should be it, thanks.

---

I've pushed the change. I chose to use an assert, which means that if it isn't a valid entry it will error out rather than falling back - which is probably preferable.

I just discovered an error however, as we choose to use the same site object, if you calculate Zhao before McVerry it will update the site object to have a siteclass based on the zhao classification. The behaviour is also similar for Rtvz if you calculate a Volcanic fault before calculating another fault (gm_hazard works around this however).

---

I haven't managed to test this rigorously and understand the siteclass enum thing yet (and if that fits with other use cases). Looks like we'll have to pick this up again next year. You're right that there's an error from the different siteclass definitions and the sequence of models being called.